wire-ios-sync-engine/Source/Use cases/GetE2eIdentityCertificatesUseCase.swift

```diff
@@ -0,0 +1,118 @@
+        self.coreCryptoProvider = coreCryptoProvider
+    }
+
+    public func invoke(mlsGroupId: MLSGroupID,
+                       clientIds: [MLSClientID]) async throws -> [E2eIdentityCertificate] {
+
+        let coreCrypto = try await coreCryptoProvider.coreCrypto(requireMLS: true)
+        let clientIds = clientIds.compactMap({$0.rawValue.data(using: .utf8)})
+        let wireIdentities = try await getWireIdentity(coreCrypto: coreCrypto,
+                                                       conversationId: mlsGroupId.data,
+                                                       clientIDs: clientIds)
+        return wireIdentities.isNonEmpty ?  try wireIdentities.compactMap({try $0.toE2eIdenityCertificate()}) : []
+    }
+
+    @MainActor
+    private func getWireIdentity(coreCrypto: SafeCoreCryptoProtocol,
+                                 conversationId: Data, clientIDs: [Data]) async throws -> [WireIdentity] {
+        return try await coreCrypto.perform {
+            return try await $0.getDeviceIdentities(
+                conversationId: conversationId,
+                deviceIds: clientIDs)
+        }
+    }
+}
+
+extension WireIdentity {
+    func toE2eIdenityCertificate() throws -> E2eIdentityCertificate? {
+        guard let certificateData = certificate.data(using: .utf8) else {
+            return nil
+        }
+        let x509Certificate = try X509Certificate(pem: certificateData)
+        return x509Certificate.toE2eIdenityCertificate(
+            clientId: clientId,
+            certificateDetails: certificate,
+            certificateStatus: status.e2eIdentityStatus,
+            mlsThumbprint: thumbprint)
+    }
+}
+
+private extension DeviceStatus {
+
+    var e2eIdentityStatus: E2EIdentityCertificateStatus {
+        switch self {
+        case .valid:
+            return .valid
+        case .expired:
+            return .expired
+        case .revoked:
+            return .revoked
+        @unknown default:
+            preconditionFailure("Should have a valid status")
+        }
+    }
+
+}
+
+extension X509Certificate {
+    func toE2eIdenityCertificate(
+        clientId: String,
+        certificateDetails: String,
+        certificateStatus: E2EIdentityCertificateStatus,
+        mlsThumbprint: String
+    ) -> E2eIdentityCertificate? {
+        guard let notValidBefore = notBefore,
+              let notValidAfter = notAfter,
+              let theSerialNumber = serialNumber?.bytes.map({String($0, radix: 16).uppercased()}).joined(separator: "")
+        else {
+            return nil
+        }
+        return .init(
+            clientId: clientId,
+            certificateDetails: certificateDetails,
+            mlsThumbprint: mlsThumbprint,
+            notValidBefore: notValidBefore,
+            expiryDate: notValidAfter,
+            certificateStatus: certificateStatus,
+            serialNumber: theSerialNumber)
+    }
+}
+
+extension Data {
```

Comment:
We have a Data extension that provides [Byte]. Can you use existing one or do you need [UInt8]?

Reply:
done